### ISSUE TITLE
using "distinct on" in postgres, is this possible?

### ISSUE THREAD

**Issue by jogibear9988** [linq2db/linq2db#2075]
is it possible to a query like here:

https://www.sisense.com/blog/first-row-per-group-5x-faster/

in oracle it seems to be possible via first_value

https://stackoverflow.com/questions/10515391/oracle-equivalent-of-postgres-distinct-on

maybe we have a extension method for example "DistinctBy<>" where we could hand over a column or a new class with multiple columns?

**Comment by sdanyliv** [comment]
Please post SQL that you need and and information about database provider.
Experiment yourself what is faster for your situation and we advise you how to create such SQL via linq2db.

**Comment by jogibear9988** [comment]
the sql is this

select distinct on (customer_id) *
from jobs
order by customer_id, priority desc, created_at

postgres now does a distinct only on customer_id, and fetches the rest of the data from the first row it gets (respecting the orderbys)

It would be nice if we‘d get a extension method „distinctby“ wich could create this on postgres
and use window functions on the others, like this:

select *
from jobs
where id in (
  select id
  from (
    select
      id,
      row_number() over (partition by customer_id
        order by priority desc, created_at) as row_num
    from jobs
  ) as ordered_jobs
  where row_num = 1
)

**Comment by sdanyliv** [comment]
I think it is "easily" to create such function even without linq2db source code change. linq2db supports window functions so we just have to create appropriate method chain expression
```cs
static IQueryable<T> DistinctBy(this IQueryable<T> source, Expression<Func<T, object>> partition, IQueryable<T> oder)
{
    // magic transformation code here
}
```
And sample usage:
```cs
var quert = 
   from j  in db.GetTable<Jobs>()
   from oj in db.GetTable<Jobs>().DistinctBy(oj => oj.CustomerId, q => q.OrderByDecending(it => it.Priority).ThenBy(it => it.CreatedAt))
                .InnerJoin(oj => oj.Id == j.Id)
   select j
```
If you have a time i can guide you how to do such transformation.

**Comment by sdanyliv** [comment]
Code inside function should dynamically create this expression:
```cs
var distinct = source.Select(q =>
   select new RowNumberHolder<T>
   {
      RN = Sql.Ext.RowNumber().Over().PartitionBy(q.CutomerId)
                    .OrderByDesc(q.Priority).ThenBy(q => q.CreatedAt).ToValue(),
      Value = q
   })
   .Where(r => r.RN == 1)
   .Select(r => r.Value);
```

**Comment by jogibear9988** [comment]
Yes, in the other dialacts we could create the method chain, but in postgres special SQL should be created!
Cause Postgres supports "distinct on" directly in SQL!

**Comment by sdanyliv** [comment]
So we have to move our time to implement additional syntax which will work only if it is a Postgres and indexes on table are correct?
FromSql is best solution for such cases.

**Comment by jogibear9988** [comment]
The synatx will also work if index are not correct ;-)

No, only wanted to know if it is possible. Maybe We could reaturn a Expression Tree and for a special provider we could (somehow magicaly via attribute) provide another method wich creates the sql

**Comment by jogibear9988** [comment]
I look if I could generate the Expression tree for the "Windows functions". Maybe you can look when I'm finish if it looks okay.

**Comment by sdanyliv** [comment]
We have to implement new method and extend SelectQuery to handle distinct fields. I do not see other ways to do that right now.

**Comment by sdanyliv** [comment]
Implement window functions variant, and then we will invent/find way how to switch implementation for specific provider. Looks like it will be good solution.

**Comment by jogibear9988** [comment]
Will do

**Comment by jogibear9988** [comment]
@sdanyliv where do I find the "RowNumberHolder<T>" ?
or do I need to implement it?

**Comment by jogibear9988** [comment]
DistinctBy is in Queryable since .Net Version 6

**Comment by jogibear9988** [comment]
As we now support DistinctBy linq method in linq2db, do we use the native SQL version in postgres?
